src/iosrv/drpc_ras.c

```diff
@@ -0,0 +1,283 @@
+/*
+ * (C) Copyright 2021 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/**
+ * This file is part of the DAOS server. It implements dRPC client RAS event
+ * functionality for communicating events to the control plane.
+ */
+
+#define D_LOGFAC DD_FAC(server)
+
+#include <daos/drpc_modules.h>
+#include <daos_srv/ras.h>
+#include <sys/utsname.h>
+#include "event.pb-c.h"
+#include "drpc_internal.h"
+
+static void
+free_ras(Shared__RASEvent *evt)
+{
+	if (evt->hostname)
+		D_FREE(evt->hostname);
+	if (evt->timestamp)
+		D_FREE(evt->timestamp);
+}
+
+static int
+init_ras(ras_event_t id, char *msg, ras_type_t type, ras_sev_t sev, char *hid,
+	 d_rank_t *rank, char *jid, uuid_t *puuid, uuid_t *cuuid,
+	 daos_obj_id_t *oid, char *cop, Shared__RASEvent *evt)
+{
+	FILE		*stream;
+	char		*buf;
+	size_t		 len;
+	struct timeval	 tv;
+	struct tm	*tm;
+	struct utsname	 uts;
+	int		 rc;
+
+	stream = open_memstream(&buf, &len);
+	if (!stream)
+		return -DER_NOMEM;
+
+	/* populate mandatory RAS fields */
+
+	evt->id = (uint32_t)id;
+	D_FPRINTF(stream, " id: [%s]", ras_event2str(id));
+
+	(void)gettimeofday(&tv, 0);
+	tm = localtime(&tv.tv_sec);
+	if (tm) {
+		D_ALLOC(evt->timestamp, DAOS_RAS_STR_FIELD_SIZE);
+		if (!evt->timestamp)
+			D_GOTO(out_fail, rc = -DER_NOMEM);
+		D_ASPRINTF(evt->timestamp,
+			   "%04d/%02d/%02d-%02d:%02d:%02d.%02ld",
+			   tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday,
+			   tm->tm_hour, tm->tm_min, tm->tm_sec,
+			   (long)tv.tv_usec / 10000);
```

Comment:
I struggle to understand

    (long)tv.tv_usec / 10000

Would strftime help hide this and the Unix Epoch (and perhaps any timezone or daylight saving issues)?

Reply:
I remember that one requirement was to have "fully qualified timestamp". I assume that this might even include specifying the timezone, no?

Reply:
I'm afraid this is just copied from @johannlombardi 's PR, happy to update or investigate further. My understanding is that storing the time zone that the event is recorded in is a specific requirement, my initial approach was to storing a time zone agnostic manner and rely on event source and consumer to handle time zone conversions.

Reply:
As i mentioned in the PR, it was just a sketch to help you. Now you took it over :)